### PR TITLE
Fix compile issues with Zcash on macOS

### PR DESCRIPTION
**Pull request by kozyilmaz**
Part of #1800.

### CODE REVIEW

File: src/crypto/equihash.cpp [hunk 1]
```diff
@@ -16,6 +16,10 @@
 #include "config/bitcoin-config.h"
 #endif
 
+#if defined(__APPLE__)
+#include "compat/endian.h"
+#endif
```
Comment (daira) [inline src/crypto/equihash.cpp]: It should be sufficient to include ``compat/endian.h`` unconditionally, I think.

Comment (daira) [inline src/crypto/equihash.cpp]: Yes, it looks like ``be32toh`` is getting accidentally included by other headers on Linux, so explicitly including ``compat/endian.h`` is better.

**Comment by kozyilmaz** [comment]
removed conditional, do you prefer another pull request with a single commit?

**Comment by daira** [comment]
@kozyilmaz:
> removed conditional, do you prefer another pull request with a single commit?

That'll do fine for now. Needs another review.

**Comment by daira** [comment]
See #2246.

**Comment by str4d** [comment]
I've removed the bug traces from the OP because Homu includes that text in the merge commit, and the individual commits already contain them. Here is the original OP text for easy reference:

---

Here are some fixes in order to compile Zcash on Apple platforms.
Some of the issues are discussed before so I just referenced them.

[macOS] “Serialize” ambiguity in wallet header, for details see https://github.com/zcash/zcash/pull/1800

```
  CXX      libbitcoin_server_a-init.o
In file included from zcash/Proof.hpp:4:0,
                 from zcash/JoinSplit.hpp:5,
                 from init.h:11,
                 from init.cpp:10:
./serialize.h: In instantiation of ‘void SerReadWrite(Stream&, const T&, int, int, CSerActionSerialize) [with Stream = CSizeComputer; T = long unsigned int]’:
wallet/wallet.h:175:9:   required from ‘void JSOutPoint::SerializationOp(Stream&, Operation, int, int) [with Stream = CSizeComputer; Operation = CSerActionSerialize]’
wallet/wallet.h:170:5:   required from here
./serialize.h:951:16: error: call of overloaded ‘Serialize(CSizeComputer&, const long unsigned int&, int&, int&)’ is ambiguous
     ::Serialize(s, obj, nType, nVersion);
     ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
```


[macOS] “compat/endian.h” is needed for __APPLE__

```
crypto/equihash.cpp: In member function ‘int Equihash<N, K>::InitialiseState(eh_HashState&)’:
crypto/equihash.cpp:33:30: error: there are no arguments to ‘htole32’ that depend on a template parameter, so a declaration of ‘htole32’ must be available [-fpermissive]
     uint32_t le_N = htole32(N);
                              ^
crypto/equihash.cpp:33:30: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
crypto/equihash.cpp:34:30: error: there are no arguments to ‘htole32’ that depend on a template parameter, so a declaration of ‘htole32’ must be available [-fpermissive]
     uint32_t le_K = htole32(K);
```


[macOS] fixes “push_back” and “Pair” ambiguities

```
  CXX      libbitcoin_server_a-rpcblockchain.o
rpcblockchain.cpp: In function ‘UniValue getblockchaininfo(const UniValue&, bool)’:
rpcblockchain.cpp:686:60: error: call of overloaded ‘Pair(const char [12], size_t)’ is ambiguous
     obj.push_back(Pair("commitments",           tree.size()));
                                                            ^
wallet/asyncrpcoperation_sendmany.cpp: In member function ‘UniValue AsyncRPCOperation_sendmany::perform_joinsplit(AsyncJoinSplitInfo&, std::vector<boost::optional<libzcash::IncrementalWitness<29ul, libzcash::SHA256Compress> > >, uint256)’:
wallet/asyncrpcoperation_sendmany.cpp:1061:42: error: call of overloaded ‘push_back(long unsigned int&)’ is ambiguous
         arrInputMap.push_back(inputMap[i]);
                                          ^
wallet/asyncrpcoperation_sendmany.cpp:1064:44: error: call of overloaded ‘push_back(long unsigned int&)’ is ambiguous
         arrOutputMap.push_back(outputMap[i]);
                                            ^
wallet/rpcwallet.cpp: In function ‘void ListTransactions(const CWalletTx&, const string&, int, bool, UniValue&, const isminefilter&)’:
wallet/rpcwallet.cpp:1339:120: error: call of overloaded ‘Pair(const char [5], size_t)’ is ambiguous
             entry.push_back(Pair("size", static_cast<CTransaction>(wtx).GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION)));
wallet/rpcwallet.cpp:1376:124: error: call of overloaded ‘Pair(const char [5], size_t)’ is ambiguous
                 entry.push_back(Pair("size", static_cast<CTransaction>(wtx).GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION)));
```

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -160,7 +160,7 @@ class JSOutPoint
     // Transaction hash
     uint256 hash;
     // Index into CTransaction.vjoinsplit
-    size_t js;
+    uint64_t js;
```
Comment (str4d) [inline src/wallet/wallet.h]: I believe we have discussed this issue before; I agree that this should be done so serialization is unambiguous, but we should check that this does indeed match existing serialization.

File: src/univalue/include/univalue.h [hunk 3]
```diff
@@ -25,6 +25,9 @@ class UniValue {
         typ = initialType;
         val = initialStr;
     }
+   UniValue(size_t val_) {
```
Comment (daira) [inline src/univalue/include/univalue.h]: Nit: indentation.

Comment (kozyilmaz) [inline src/univalue/include/univalue.h]: @daira I created a pull request to @str4d UniValue tree, once it's accepted I'll merge them in, so this indentation issue won't be present.

File: src/wallet/wallet.h (hunk 2 above)
Comment (kozyilmaz) [inline src/wallet/wallet.h]: @str4d After pull request to your subtree is in, do you want me to create another PR?

**Comment by str4d** [comment]
@kozyilmaz I've merged your PR into my branch. Now the corresponding commit should be removed from this branch, and then the UniValue branch pulled in as a git subtree. If you have problems with this, I can do so.

**Comment by kozyilmaz** [comment]
Removed univalue commit from this PR however I failed to pull univalue subtree (I am not so familiar with git subtree merges), can you write down correct git commands
```
bash-3.2$ git fetch https://github.com/str4d/univalue.git size_t
warning: no common commits
remote: Counting objects: 492, done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 492 (delta 0), reused 0 (delta 0), pack-reused 486
Receiving objects: 100% (492/492), 86.99 KiB | 0 bytes/s, done.
Resolving deltas: 100% (261/261), done.
From https://github.com/str4d/univalue
 * branch            size_t     -> FETCH_HEAD

bash-3.2$ ./contrib/devtools/git-subtree-check.sh src/univalue
src/univalue in HEAD was last updated to upstream commit 9ef5b78c1998509b8f1ccd76f0aee15140e384be (tree 9081e61911ac408916841054e7085a3843440f66)
src/univalue in HEAD currently refers to tree 9081e61911ac408916841054e7085a3843440f66
GOOD

bash-3.2$ git subtree pull --prefix src/univalue https://github.com/str4d/univalue.git size_t
From https://github.com/str4d/univalue
 * branch            size_t     -> FETCH_HEAD
fatal: refusing to merge unrelated histories
```

**Comment by kozyilmaz** [comment]
@str4d synced subtree with the commands below, let me if that's ok

```
$ git remote add -f univalue https://github.com/str4d/univalue.git
$ git checkout master
$ git subtree pull --prefix=src/univalue/ univalue size_t --squash
```

seems to be alright.
```
$ ./contrib/devtools/git-subtree-check.sh src/univalue
src/univalue in HEAD was last updated to upstream commit 651aba64cc2a1882bcd56f006bcf2736109fe16c (tree 7e3a80ca51fde0c1deefc5d8d68fda069651ff40)
src/univalue in HEAD currently refers to tree 7e3a80ca51fde0c1deefc5d8d68fda069651ff40
GOOD
```

**Comment by kozyilmaz** [comment]
ping @str4d @daira 
In order to complete Zcash compilation on macOS, trivial libsnark patches (usage of ar script) are also needed but I think you also prefer to keep it in a different PR due to https://github.com/zcash/zcash/issues/2252

**Comment by nathan-at-least** [comment]
I've bumped this from 1.0.9. Before accepting patches for other platforms, we need to integrate builders for those platforms into our CI environment.

**Comment by kozyilmaz** [comment]
@nathan-at-least Understandable, may I ask what Zcash team is planning for macOS CI machine? I have these questions so if there is a clear plan maybe I can help too:
1. What will be the toolchain (native macOS clang vs latest (5.x or 6.x) gnu gcc built on macOS)?
* libsnark (and successor libff) does not compile with macOS clang and probably won't be supported at all.
* How about https://github.com/ebfull/bellman, is this considered as a libsnark replacement? If so clang may be on the table again.
2. If precompiled gnu gcc on macOS will be the toolchain, are you planning to use homebrew? I got a build system that compiles every bit of software needed (including gcc) but it is probably not suitable for a CI system that checks every commit (https://github.com/kozyilmaz/zcash-apple)

**Comment by nathan-at-least** [comment]
This is now blocked on #2469.

**Comment by kozyilmaz** [comment]
https://github.com/zcash/zcash/pull/2697 already added macOS patches, closing for now
